app/uk/gov/hmrc/gform/validation/TaxPeriodDateValidator.scala

```diff
@@ -0,0 +1,115 @@
+import uk.gov.hmrc.gform.validation.ValidationServiceHelper.{ validationFailure, validationSuccess }
+import uk.gov.hmrc.gform.validation.ValidationUtil.ValidatedType
+
+class TaxPeriodDateValidator[D <: DataOrigin](formModelVisibilityOptics: FormModelVisibilityOptics[D])(implicit
+  messages: Messages,
+  sse: SmartStringEvaluator
+) {
+
+  def validate(formComponent: FormComponent): ValidatedType[Unit] =
+    validateRequired(formComponent)
+      .andThen(_ =>
+        validateMonthYear(formComponent)
+          .andThen(validateMonthYearCombo(formComponent, _))
+      )
+
+  def validateMonthYear(formComponent: FormComponent): ValidatedType[(Int, Int)] =
+    formComponent.multiValueId.atomsModelComponentIds.map(formModelVisibilityOptics.data.one) match {
+      case Some(month) :: Some(year) :: Nil =>
+        val label = fieldDescriptor(formComponent, "")
+
+        val yearValidated = hasMaximumLength(year, 4, label + " " + messages("date.year"))
+          .andThen(_ => isNumeric(year, label + " " + messages("date.year"), label))
+          .leftMap(error => Map(errorGranularity(formComponent)(TaxPeriodDate.year) -> Set(error)))
+
+        val monthValidated = hasMaximumLength(month, 2, label + " " + messages("date.month"))
+          .andThen(_ => isNumeric(month, label + " " + messages("date.month"), label))
+          .leftMap(error => Map(errorGranularity(formComponent)(TaxPeriodDate.month) -> Set(error)))
+
+        (monthValidated, yearValidated).mapN((month, year) => (month, year))
+      case _ =>
+        validationFailure(formComponent.firstAtomModelComponentId, formComponent, "date.isMissing", None, "")
+    }
+
+  private def validateMonthYearCombo(formComponent: FormComponent, monthYear: (Int, Int)): ValidatedType[Unit] = {
+    val label = fieldDescriptor(formComponent, "")
+    val monthLabel = label + " " + messages("date.month")
+    val yearLabel = label + " " + messages("date.year")
+    val (month, year) = monthYear
+    validationSuccess
+      .ensure(
+        Map(
+          errorGranularity(formComponent)(TaxPeriodDate.month) -> Set(
+            messages("generic.error.mustBeBetween", monthLabel, 1, 12)
+          )
+        )
+      )(_ => month >= 1 && month <= 12)
+      .ensure(
+        Map(
+          errorGranularity(formComponent)(TaxPeriodDate.year) -> Set(
+            messages("generic.error.mustBeBetween", yearLabel, 1900, 2099)
+          )
+        )
+      )(_ => year >= 1900 && year <= 2099)
+  }
+
+  private def validateRequired(
+    formComponent: FormComponent
+  ): ValidatedType[Unit] = {
+
+    case class ModelComponentIdValue(modelComponentId: ModelComponentId, value: Option[String])
+
+    val atomsWithValues: List[ModelComponentIdValue] = formComponent.multiValueId.atomsModelComponentIds.map(m =>
+      ModelComponentIdValue(m, formModelVisibilityOptics.data.one(m))
+    )
+
+    val validatedResult = if (formComponent.mandatory) {
+      atomsWithValues.map { mcv =>
```

Comment:
This `map` can be changed to `foldMap` so the results of the lambda are immediately combined with theirs `Monoid` instance.